graphql.go

```diff
@@ -0,0 +1,149 @@
+package goshopify
+
+import (
+	"math"
+	"time"
+)
+
+// GraphQLService is an interface to interact with the graphql endpoint
+// of the Shopify API
+// See https://shopify.dev/docs/admin-api/graphql/reference
+type GraphQLService interface {
+	Query(string, interface{}, interface{}) (int, error)
+}
+
+// GraphQLServiceOp handles communication with the graphql endpoint of
+// the Shopify API.
+type GraphQLServiceOp struct {
+	client *Client
+}
+
+type graphQLResponse struct {
+	Data       interface{}       `json:"data"`
+	Errors     []graphQLError    `json:"errors"`
+	Extensions *graphQLExtension `json:"extensions"`
+}
+
+type graphQLExtension struct {
+	Cost GraphQLCost `json:"cost"`
+}
+
+// GraphQLCost represents the cost of the graphql query
+type GraphQLCost struct {
+	RequestedQueryCost int                   `json:"requestedQueryCost"`
+	ActualQueryCost    *int                  `json:"actualQueryCost"`
+	ThrottleStatus     GraphQLThrottleStatus `json:"throttleStatus"`
+}
+
+// GraphQLThrottleStatus represents the status of the shop's rate limit points
+type GraphQLThrottleStatus struct {
+	MaximumAvailable   float64 `json:"maximumAvailable"`
+	CurrentlyAvailable float64 `json:"currentlyAvailable"`
+	RestoreRate        float64 `json:"restoreRate"`
+}
+
+type graphQLError struct {
+	Message    string                  `json:"message"`
+	Extensions *graphQLErrorExtensions `json:"extensions"`
+	Locations  []graphQLErrorLocation  `json:"locations"`
+}
+
+type graphQLErrorExtensions struct {
+	Code          string
+	Documentation string
+}
+
+const (
+	graphQLErrorCodeThrottled = "THROTTLED"
+)
+
+type graphQLErrorLocation struct {
+	Line   int `json:"line"`
+	Column int `json:"column"`
+}
+
+// Query creates a graphql query against the Shopify API
+// the "data" portion of the response is unmarshalled into resp
+// Returns the number of attempts required to perform the request
```

Comment:
I think it's worth mentioning that `Query()` will retry throttled requests but not e..g any other type of 500 error.

Reply:
To keep the concept of retry more defined, I added this comment to the `WithRetry` method

```golang
// WithRetry set the number of times a request will be retried if a rate limit or service unavailable error is returned.
// Rate limiting can be either REST API limits or GraphQL Cost limits
```